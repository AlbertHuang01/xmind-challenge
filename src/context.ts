import { queryBillList, addBillItem, queryCategoryList } from './service';
import { Bill, BILL_TYPE, BillCategory } from './model';
import { createContext, useEffect, useMemo, useState, useContext } from "react";
import dayjs from 'dayjs';

interface Condition {
  date?: {
    year: number;
    month: number;
  },
  categoryId?: string
}

const context = createContext({})

export const Provicer = context.Provider

export function initContext() {
  const [billList, setBillList] = useState<Bill[]>([])
  const [categoryList, setCategoryList] = useState<BillCategory[]>([])
  const [condition, setCondition] = useState<Condition>(null)
  const [addBillVisible,setAddBillVisible]=useState(false)

  const billListFilted = useMemo(() => {
    let result = billList;
    if (condition) {
      const { date, categoryId } = condition
      if (date) {
        const { year, month } = date
        result = result.filter(bill => {
          const time = dayjs(bill.time)
          return time.year() === year && time.month() + 1 === month
        })
      }
      if (categoryId) {
        result = result.filter(bill => bill.category === categoryId)
      }
    }
    return result
  }, [billList, condition])

  const billListGroupByType = useMemo(() => {
    const income = billListFilted.filter(bill => bill.type === BILL_TYPE.INCOME).reduce((prev, curr) => prev + curr.amount, 0)
    const expenditure = billListFilted.filter(bill => bill.type === BILL_TYPE.EXPENDITURE).reduce((prev, curr) => prev + curr.amount, 0)
    return { income, expenditure }
  }, [billListFilted])

  const addBill = (bill: Bill) => {
    addBillItem(bill).then(() => {
      queryBillList().then(setBillList)
    })
  }

  useEffect(() => {
    queryBillList().then(setBillList)
    queryCategoryList().then(setCategoryList)
  }, [])

  return {
    billList: billListFilted,
    billListGroupByType,
    condition,
    setCondition,
    addBill,
    categoryList,
    addBillVisible,
    setAddBillVisible
  }
}

export const useAppContext = () => {
  return useContext(context) as ReturnType<typeof initContext>
}
