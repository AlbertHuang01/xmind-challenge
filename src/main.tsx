import { useEffect } from "react";
import axios from "axios";

import SearchForm from "./search-form";
import zhCN from "antd/lib/locale/zh_CN";
import { Card, ConfigProvider, Space } from "antd";
import "./main.css";
import BillList from "./bill-list";
import AddBill from "./add-bill";
import { GroupByCategoryType, GroupByIncomeAndExpenditure } from "./bill-list-group";
import { appContextInit, APP_DATA_CONTEXT } from "./context";

// 设置 axios 请求时的 baseURL
axios.defaults.baseURL = "http://127.0.0.1:3000";

export const APP_DATA = APP_DATA_CONTEXT

export function App() {
  const appContext=appContextInit()

  return (
    <ConfigProvider locale={zhCN}>
      <Card style={{ width: "80%", margin: "40px auto" }}>
        <APP_DATA.Provider
          value={appContext}
        >
          <Space direction={"vertical"} style={{ display: "flex" }}>
            <SearchForm />
            <BillList />
          </Space>
          <AddBill />
          <GroupByIncomeAndExpenditure />
          <GroupByCategoryType />
        </APP_DATA.Provider>
      </Card>
    </ConfigProvider>
  );
}
