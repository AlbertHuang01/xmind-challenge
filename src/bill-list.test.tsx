import { render } from "@testing-library/react"
import { renderHook } from "@testing-library/react-hooks"
import BillList from "./bill-list"
import { appContextInit } from "./context"
import { APP_DATA } from "./main"

describe('bill-list',()=>{
  it('bill-list snapshot',async()=>{
    const {waitForNextUpdate,result}=renderHook(()=>appContextInit())
    await waitForNextUpdate()

    const wrapper = ({ children }) => <APP_DATA.Provider value={result.current}>{children}</APP_DATA.Provider>
    const {baseElement}=render(<BillList/>,{wrapper})
    expect(baseElement.querySelector('.ant-table-tbody')).toMatchSnapshot()
  })
})
