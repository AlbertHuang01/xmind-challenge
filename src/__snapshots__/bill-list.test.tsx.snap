// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bill-list bill-list snapshot 1`] = `
<tbody
  class="ant-table-tbody"
>
  <tr
    class="ant-table-row ant-table-row-level-0"
    data-row-key="1561910400000-0-8s0p77c323-5400"
  >
    <td
      class="ant-table-cell"
    >
      2019-07-01 00:00:00
    </td>
    <td
      class="ant-table-cell"
    >
      支出
    </td>
    <td
      class="ant-table-cell"
    >
      房贷
    </td>
    <td
      class="ant-table-cell"
    >
      5400
    </td>
  </tr>
  <tr
    class="ant-table-row ant-table-row-level-0"
    data-row-key="1561910400000-0-0fnhbcle6hg-1500"
  >
    <td
      class="ant-table-cell"
    >
      2019-07-01 00:00:00
    </td>
    <td
      class="ant-table-cell"
    >
      支出
    </td>
    <td
      class="ant-table-cell"
    />
    <td
      class="ant-table-cell"
    >
      1500
    </td>
  </tr>
  <tr
    class="ant-table-row ant-table-row-level-0"
    data-row-key="1563897600000-0-3tqndrjqgrg-3900"
  >
    <td
      class="ant-table-cell"
    >
      2019-07-24 00:00:00
    </td>
    <td
      class="ant-table-cell"
    >
      支出
    </td>
    <td
      class="ant-table-cell"
    />
    <td
      class="ant-table-cell"
    >
      3900
    </td>
  </tr>
  <tr
    class="ant-table-row ant-table-row-level-0"
    data-row-key="1564502400000-0-bsn20th0k2o-1900"
  >
    <td
      class="ant-table-cell"
    >
      2019-07-31 00:00:00
    </td>
    <td
      class="ant-table-cell"
    >
      支出
    </td>
    <td
      class="ant-table-cell"
    />
    <td
      class="ant-table-cell"
    >
      1900
    </td>
  </tr>
</tbody>
`;
